app/controllers/fhir/r4/api_controller.rb

```diff
@@ -342,12 +338,106 @@ def create
                             created_by: @current_actor_label,
                             comment: "New vaccine added via API (ID: #{resource.id}).")
       end
+    when 'observation'
+      return if doorkeeper_authorize!(*OBSERVATION_WRITE_SCOPES)
+
+      resource, fhir_map = build_laboratory(contents)
+
+      ActiveRecord::Base.transaction do
+        unless referenced_patient_valid_for_client?(resource, :patient_id) && resource.save(context: :api) && fhir_map.all? { |_k, v| v[:errors].blank? }
+          req_json = JSON.parse(request_body)
+          status_unprocessable_entity(resource, fhir_map, req_json) && (raise SaveError)
+        end
+
+        Rails.logger.info "Created Lab Result (ID: #{resource.id}) for Patient with ID: #{resource.patient_id}"
+        History.lab_result(patient: resource.patient_id,
+                           created_by: @current_actor_label,
+                           comment: "New lab result added via API (ID: #{resource.id}).")
+      end
     else
       status_not_found && return
     end
     status_created(resource.as_fhir) && return
   rescue JSON::ParserError
     status_bad_request(['Invalid JSON in request body'])
+  rescue SaveError
+    nil
+  end
+
+  # Create a set of resources as an atomic action
+  #
+  # Supports (writing): Patient, Observation
+  #
+  # POST /fhir/r4
+  def transaction
+    status_unsupported_media_type && return unless content_type_header?('application/fhir+json')
+
+    # Must have Patient write scopes
+    return if doorkeeper_authorize!(*PATIENT_WRITE_SCOPES)
+
+    # Parse in the FHIR
+    request_body = request.body.read
+    contents = FHIR.from_contents(request_body) unless request_body.blank?
+    errors = contents&.validate
```

Comment:
The check for the max of 50 entries should happen between these two lines, and then return if it's over the threshold. I haven't dug into it, but ostensibly, the `validate` call is really slow (seeing ~10 seconds for bundle with 1170 patients) so better to just skip it when we can.

Reply:
Good point, updated in 37df199.